kvm-ioctls/src/ioctls/vcpu.rs

```diff
@@ -2104,6 +2106,65 @@ impl VcpuFd {
         }
     }
 
+    /// Maps the coalesced MMIO ring page. This allows reading entries from
+    /// the ring via [`coalesced_mmio_read()`](VcpuFd::coalesced_mmio_read).
+    ///
+    /// # Returns
+    ///
+    /// Returns an error if the buffer could not be mapped, usually either because
+    /// `KVM_CAP_DIRTY_LOG_RING` ([`Cap::DirtyLogRing`](crate::Cap::DirtyLogRing))
+    /// is not available.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # extern crate kvm_ioctls;
+    /// # use kvm_ioctls::{Cap, Kvm};
+    /// let kvm = Kvm::new().unwrap();
+    /// let vm = kvm.create_vm().unwrap();
+    /// let mut vcpu = vm.create_vcpu(0).unwrap();
+    /// if kvm.check_extension(Cap::DirtyLogRing) {
+    ///     vcpu.coalesced_mmio_ring().unwrap();
+    /// }
+    /// ```
+    pub fn map_dirty_log_ring(&mut self, bytes: usize) -> Result<()> {
+        if self.dirty_log_ring.is_none() {
+            let ring = KvmDirtyLogRing::mmap_from_fd(&self.vcpu, bytes)?;
+            self.dirty_log_ring = Some(ring);
+        }
+        Ok(())
+    }
+
+    /// Gets the dirty log ring iterator if one is mapped.
+    ///
+    /// Returns an iterator over dirty guest frame numbers as (slot, offset) tuples.
+    /// Returns `None` if no dirty log ring has been mapped via [`map_dirty_log_ring`](VcpuFd::map_dirty_log_ring).
+    ///
+    /// # Returns
+    ///
+    /// An optional iterator over the dirty log ring entries.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// # use kvm_ioctls::Kvm;
+    /// # use kvm_ioctls::Cap;
+    /// let kvm = Kvm::new().unwrap();
+    /// let vm = kvm.create_vm().unwrap();
+    /// let mut vcpu = vm.create_vcpu(0).unwrap();
+    /// if kvm.check_extension(Cap::DirtyLogRing) {
+    ///     vcpu.map_dirty_log_ring(4096).unwrap();
+    ///     if let Some(mut iter) = vcpu.dirty_log_ring_iter() {
+    ///         for (slot, offset) in iter {
+    ///             println!("Dirty page in slot {} at offset {}", slot, offset);
+    ///         }
+    ///     }
+    /// }
+    /// ```
+    pub fn dirty_log_ring_iter(&mut self) -> Option<&mut KvmDirtyLogRing> {
```

Comment:
we can hide the actual iterator type here by just returning `impl Iterator<Item = blablabla>`. Then maybe the entire struct doesnt need to be exported?